core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java

```diff
@@ -0,0 +1,204 @@
+            parsed = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+
+        assertNotNull(parsed);
+        assertThat(parsed.getMessage(), equalTo("Elasticsearch exception [type=foo, reason=something went wrong]"));
+    }
+
+    public void testFromXContentWithCause() throws IOException {
+        ElasticsearchException e = new ElasticsearchException("foo",
+                new ElasticsearchException("bar",
+                        new ElasticsearchException("baz",
+                                new RoutingMissingException("_test", "_type", "_id"))));
+
+        final XContent xContent = randomFrom(XContentType.values()).xContent();
+        XContentBuilder builder = XContentBuilder.builder(xContent).startObject().value(e).endObject();
+
+        ElasticsearchException parsed;
+        try (XContentParser parser = xContent.createParser(builder.bytes())) {
+            assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+            parsed = ElasticsearchException.fromXContent(parser);
+            assertEquals(XContentParser.Token.END_OBJECT, parser.currentToken());
+            assertNull(parser.nextToken());
+        }
+
+        assertNotNull(parsed);
+        assertThat(parsed.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=foo]"));
+
+        ElasticsearchException cause = (ElasticsearchException) parsed.getCause();
+        assertThat(cause.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=bar]"));
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertThat(cause.getMessage(), equalTo("Elasticsearch exception [type=exception, reason=baz]"));
+
+        cause = (ElasticsearchException) cause.getCause();
+        assertThat(cause.getMessage(),
+                equalTo("Elasticsearch exception [type=routing_missing_exception, reason=routing is required for [_test]/[_type]/[_id]]"));
+        assertThat(cause.getHeaderKeys(), hasSize(2));
+        assertThat(cause.getHeader("index").size(), is(1));
+        assertThat(cause.getHeader("index").get(0), equalTo("_test"));
+        assertThat(cause.getHeader("index_uuid").size(), is(1));
+        assertThat(cause.getHeader("index_uuid").get(0), equalTo("_na_"));
+    }
+
+    private static void assertToXContentAsJson(ToXContent toXContent, boolean skipTrace, Supplier<Matcher<String>> expected)
+            throws IOException {
+        ToXContent.Params params = ToXContent.EMPTY_PARAMS;
+        if (skipTrace == false) {
+            params = new ToXContent.MapParams(Collections.singletonMap(ElasticsearchException.REST_EXCEPTION_SKIP_STACK_TRACE, "false"));
+        }
+        try (XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent())) {
+            builder.startObject();
```

Comment:
as I said above, it would be nice not to require this, but start and end within toXContent. Not sure how feasible it is.